mlir/lib/Dialect/Math/Transforms/ExpandPatterns.cpp

```diff
@@ -311,40 +311,83 @@ static LogicalResult convertFPowIOp(math::FPowIOp op,
   return success();
 }
 
-// Converts  Powf(float a, float b) (meaning a^b) to exp^(b * ln(a))
+// Convert Powf(float a, float b) for special cases when b is constant:
+// when b == 0, or |b| == 0.5, 1.0, or 2.0.
+static LogicalResult convertSpecialPowfOp(math::PowFOp op,
+                                          PatternRewriter &rewriter) {
+  ImplicitLocOpBuilder b(op->getLoc(), rewriter);
+  Value operandA = op.getOperand(0);
+  Value operandB = op.getOperand(1);
+  auto opType = operandA.getType();
+  auto baseType = operandB.getType();
+
+  auto &sem = dyn_cast<mlir::FloatType>(getElementTypeOrSelf(baseType))
+                  .getFloatSemantics();
+
+  auto valueB = APFloat(sem);
+  if (!matchPattern(operandB, m_ConstantFloat(&valueB))) {
+    // Not a constant, return failure
+    return failure();
+  }
+  float floatValueB = valueB.convertToFloat();
```

Comment:
Avoid converting to C++ `float`, as the actual type could have higher precision, so this conversion would be rounding to lesser precision and could end up enabling an incorrect rewrite, e.g. if the type is `f64`, then the rewrite from `powf(a, 1.0 + 1.0e-10)` into `a` would be incorrect as, in `f64`, 1.0 + 1.0e-10 != 1.0`, but the rounded `floatValueB` is exactly `1.0`.

Can you keep the whole logic in this function in APFloat?

Reply:
Done, but this is might be bit verbose so could you check it please?

Reply:
@hanhanW , i am not familiar myself with APFloat, can you review that aspect?

Reply:
I'm not very familiar with it either. After skimming through the doc, I think we can use `isExactlyValue` method. I think we do not have precision issue for `0`, `+-1`, `+-0.5`, `+-2` numbers. 

https://github.com/llvm/llvm-project/blob/bee9664970d51df3f4e1d298d1bcb95bba364e17/llvm/include/llvm/ADT/APFloat.h#L1420-L1433

Reply:
Fixed!